### ISSUE TITLE
The `Error` DSL has some problems

### ISSUE THREAD

**Issue by ikawaha** [goadesign/goa#2867]
The `Error` DSL seems to have a few problems.

# TL;DR;

1. API level `Error` generates invalid code if `ResultType` is set.
    - only references are allowed(https://github.com/goadesign/goa/pull/2555).
2. If error responses with the same name but different responses are specified in `Service` and `Method`, invalid code will be generated.
3. When `Error`s of the same name with different `ResultType` are defined in Service and Method, only the error defined as Service will be valid.

# Case 1

API level `Error` generates invalid code if `ResultType` is set.

-> PR:https://github.com/goadesign/goa/pull/2863

## Design example

```go
package design

import (
	. "goa.design/goa/v3/dsl"
)

var _ = API("calc", func() {
	Error("DivByZero", String)
	HTTP(func() {
		Response("DivByZero", StatusBadRequest)
	})
})

var _ = Service("calc", func() {
	Method("div", func() {
		Error("DivByZero")
		Payload(func() {
			Field(1, "a", Int, "Left operand")
			Field(2, "b", Int, "Right operand")
			Required("a", "b")
		})
		Result(Int)
		HTTP(func() {
			GET("/div/{a}/{b}")
		})
	})
})
```

## Generated code

```go: gen/http/server/encode_decode.go
func EncodeDivError(encoder func(context.Context, http.ResponseWriter) goahttp.Encoder, formatter func(err error) goahttp.Statuser) func(context.Context, http.ResponseWriter, error) error {
	encodeError := goahttp.ErrorEncoder(encoder, formatter)
	return func(ctx context.Context, w http.ResponseWriter, v error) error {
		en, ok := v.(ErrorNamer)
		if !ok {
			return encodeError(ctx, w, v)
		}
		switch en.ErrorName() {
		case "DivByZero":
			res := v.(string)        // ← Impossible type assertion: `string` does not implement 'error'
			enc := encoder(ctx, w)
			body := res
			w.Header().Set("goa-error", res.ErrorName())
			w.WriteHeader(http.StatusBadRequest)
			return enc.Encode(body)
		default:
			return encodeError(ctx, w, v)
		}
	}
}

```

# Case 2

If error responses with the same name but different responses are specified in `Service` and `Method`, invalid code will be generated.

-> PR:https://github.com/goadesign/goa/pull/2865

## Design example

```go
package design

import (
	. "goa.design/goa/v3/dsl"
)

var _ = Service("calc", func() {
	Error("DivByZero")
	HTTP(func() {
		Response("DivByZero", StatusInternalServerError)
	})
	Method("div", func() {
		Payload(func() {
			Field(1, "a", Int, "Left operand")
			Field(2, "b", Int, "Right operand")
			Required("a", "b")
		})
		Result(Int)
		HTTP(func() {
			GET("/div/{a}/{b}")
			Response("DivByZero", StatusBadRequest)
		})
	})
})
```

### Generated code

```go
func EncodeDivError(encoder func(context.Context, http.ResponseWriter) goahttp.Encoder, formatter func(err error) goahttp.Statuser) func(context.Context, http.ResponseWriter, error) error {
	encodeError := goahttp.ErrorEncoder(encoder, formatter)
	return func(ctx context.Context, w http.ResponseWriter, v error) error {
		en, ok := v.(ErrorNamer)
		if !ok {
			return encodeError(ctx, w, v)
		}
		switch en.ErrorName() {
		case "DivByZero":                  // ← ★
			res := v.(*goa.ServiceError)
			enc := encoder(ctx, w)
			var body interface{}
			if formatter != nil {
				body = formatter(res)
			} else {
				body = NewDivDivByZeroResponseBody(res)
			}
			w.Header().Set("goa-error", res.ErrorName())
			w.WriteHeader(http.StatusBadRequest)
			return enc.Encode(body)
		case "DivByZero":                   // ← ★
			res := v.(*goa.ServiceError)
			enc := encoder(ctx, w)
			var body interface{}
			if formatter != nil {
				body = formatter(res)
			} else {
				body = NewDivDivByZeroResponseBody(res)
			}
			w.Header().Set("goa-error", res.ErrorName())
			w.WriteHeader(http.StatusInternalServerError)
			return enc.Encode(body)
		default:
			return encodeError(ctx, w, v)
		}
	}
}
```

# Case 3

When `Error`s of the same name with different `ResultType` are defined in Service and Method, only the error defined as Service will be valid.

-> PR: https://github.com/goadesign/goa/pull/2866

# Design example

```go
package design

import (
	. "goa.design/goa/v3/dsl"
)

var _ = Service("calc", func() {
	Error("DivByZero", String)
	HTTP(func() {
		Response("DivByZero", StatusInternalServerError)
	})
	Method("div", func() {
		Error("DivByZero", Int)
		Payload(func() {
			Field(1, "a", Int, "Left operand")
			Field(2, "b", Int, "Right operand")
			Required("a", "b")
		})
		Result(Int)
		HTTP(func() {
			GET("/div/{a}/{b}")
			Response("DivByZero", StatusBadRequest)
		})
	})
})
```

# Generated code

```go
type DivByZero string // ← defined in service.

// Error returns an error description.
func (e DivByZero) Error() string {
	return ""
}

// ErrorName returns "DivByZero".
func (e DivByZero) ErrorName() string {
	return "DivByZero"
}
```
